From 7649219b6916fb040beaf8a3c8217cdfb545d754 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Thu, 5 Sep 2013 13:50:49 +0100
Subject: [PATCH 075/351] ASoC: wm5102: Factor out patch application

In preparation for additional steps in the patch application process.

Change-Id: I8a8c2c2b40e15653da8a07ea45fa7e4b1ce278c2
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 drivers/mfd/wm5102-tables.c      |   33 ++++++++++++++++++++++-----------
 include/linux/mfd/arizona/core.h |    3 +++
 sound/soc/codecs/wm5102.c        |    7 ++-----
 3 files changed, 27 insertions(+), 16 deletions(-)

diff --git a/drivers/mfd/wm5102-tables.c b/drivers/mfd/wm5102-tables.c
index 0755244..52a71b7 100644
--- a/drivers/mfd/wm5102-tables.c
+++ b/drivers/mfd/wm5102-tables.c
@@ -79,12 +79,32 @@ static const struct reg_default wm5102_revb_patch[] = {
 	{ 0x470, 0xC01 },
 };
 
+int wm5102_apply_patch(struct arizona *arizona,
+		       const struct reg_default *wm5102_patch,
+		       const int patch_size)
+{
+	int i, ret;
+
+	for (i = 0; i < patch_size; i++) {
+		ret = regmap_write(arizona->regmap, wm5102_patch[i].reg,
+				   wm5102_patch[i].def);
+		if (ret != 0) {
+			dev_err(arizona->dev, "Failed to write %x = %x: %d\n",
+				wm5102_patch[i].reg, wm5102_patch[i].def, ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(wm5102_apply_patch);
+
 /* We use a function so we can use ARRAY_SIZE() */
 int wm5102_patch(struct arizona *arizona)
 {
 	const struct reg_default *wm5102_patch;
 	int ret = 0;
-	int i, patch_size;
+	int patch_size;
 
 	switch (arizona->rev) {
 	case 0:
@@ -99,16 +119,7 @@ int wm5102_patch(struct arizona *arizona)
 
 	regcache_cache_bypass(arizona->regmap, true);
 
-	for (i = 0; i < patch_size; i++) {
-		ret = regmap_write(arizona->regmap, wm5102_patch[i].reg,
-				   wm5102_patch[i].def);
-		if (ret != 0) {
-			dev_err(arizona->dev, "Failed to write %x = %x: %d\n",
-				wm5102_patch[i].reg, wm5102_patch[i].def, ret);
-			goto out;
-		}
-	}
-
+	ret = wm5102_apply_patch(arizona, wm5102_patch, patch_size);
 out:
 	regcache_cache_bypass(arizona->regmap, false);
 	return ret;
diff --git a/include/linux/mfd/arizona/core.h b/include/linux/mfd/arizona/core.h
index 06a52bc..c9777a4 100644
--- a/include/linux/mfd/arizona/core.h
+++ b/include/linux/mfd/arizona/core.h
@@ -121,6 +121,9 @@ int arizona_request_irq(struct arizona *arizona, int irq, char *name,
 void arizona_free_irq(struct arizona *arizona, int irq, void *data);
 int arizona_set_irq_wake(struct arizona *arizona, int irq, int on);
 
+extern int wm5102_apply_patch(struct arizona *arizona,
+			      const struct reg_default *wm5102_patch,
+			      const int patch_size);
 int wm5102_patch(struct arizona *arizona);
 int florida_patch(struct arizona *arizona);
 int wm8997_patch(struct arizona *arizona);
diff --git a/sound/soc/codecs/wm5102.c b/sound/soc/codecs/wm5102.c
index d7f7931..da104e0 100644
--- a/sound/soc/codecs/wm5102.c
+++ b/sound/soc/codecs/wm5102.c
@@ -585,9 +585,8 @@ static int wm5102_sysclk_ev(struct snd_soc_dapm_widget *w,
 {
 	struct snd_soc_codec *codec = w->codec;
 	struct arizona *arizona = dev_get_drvdata(codec->dev->parent);
-	struct regmap *regmap = codec->control_data;
 	const struct reg_default *patch = NULL;
-	int i, patch_size;
+	int patch_size;
 
 	switch (arizona->rev) {
 	case 0:
@@ -603,9 +602,7 @@ static int wm5102_sysclk_ev(struct snd_soc_dapm_widget *w,
 	switch (event) {
 	case SND_SOC_DAPM_POST_PMU:
 		if (patch)
-			for (i = 0; i < patch_size; i++)
-				regmap_write_async(regmap, patch[i].reg,
-						   patch[i].def);
+			wm5102_apply_patch(arizona, patch, patch_size);
 		break;
 
 	default:
-- 
1.7.9.5

